atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/QosCassandraReadEteTest.java

```diff
@@ -0,0 +1,147 @@
+import com.palantir.atlasdb.keyvalue.api.RowResult;
+import com.palantir.atlasdb.qos.ratelimit.RateLimitExceededException;
+import com.palantir.atlasdb.table.description.ValueType;
+import com.palantir.atlasdb.todo.ImmutableTodo;
+import com.palantir.atlasdb.todo.Todo;
+import com.palantir.atlasdb.todo.TodoSchema;
+import com.palantir.common.base.BatchingVisitable;
+
+public class QosCassandraReadEteTest extends QosCassandraEteTestSetup {
+    private static final int ONE_TODO_SIZE_IN_BYTES = 1050;
+
+    @BeforeClass
+    public static void createTransactionManagerAndWriteData() {
+        ensureTransactionManagerIsCreated();
+        writeNTodosOfSize(200, 1_000);
+        serializableTransactionManager.close();
+    }
+
+    @Test
+    public void shouldBeAbleToReadSmallAmountOfBytesIfDoesNotExceedLimit() {
+        assertThat(readOneBatchOfSize(1)).hasSize(1);
+    }
+
+    @Test
+    public void shouldBeAbleToReadSmallAmountOfBytesSeriallyIfDoesNotExceedLimit() {
+        IntStream.range(0, 50).forEach(i -> assertThat(readOneBatchOfSize(1)).hasSize(1));
+    }
+
+    @Test
+    public void shouldBeAbleToReadLargeAmountsExceedingTheLimitFirstTime() {
+        assertThat(readOneBatchOfSize(12)).hasSize(12);
+    }
+
+    @Test
+    public void shouldBeAbleToReadLargeAmountsExceedingTheLimitSecondTimeWithSoftLimiting() {
+        assertThat(readOneBatchOfSize(12)).hasSize(12);
+        // The second read might actually be faster as the transaction/metadata is cached
+        assertThat(readOneBatchOfSize(12)).hasSize(12);
+    }
+
+    @Test
+    public void shouldNotBeAbleToReadLargeAmountsIfSoftLimitSleepWillBeMoreThanConfiguredBackoffTime() {
+        assertThatThrownBy(() -> readOneBatchOfSize(200))
+                .isInstanceOf(RateLimitExceededException.class)
+                .hasMessage("Rate limited. Available capacity has been exhausted.");
+    }
+
+    @Test
+    public void readRateLimitShouldBeRespectedByConcurrentReadingThreads() throws InterruptedException {
+        int numThreads = 5;
+        int numReadsPerThread = 10;
+        ExecutorService executorService = Executors.newFixedThreadPool(numThreads);
+        List<Future<List<Todo>>> futures = new ArrayList<>(numThreads);
+
+        long start = System.nanoTime();
+        IntStream.range(0, numThreads).forEach(i ->
+                futures.add(executorService.submit(() -> {
+                    List<Todo> results = new ArrayList<>(numReadsPerThread);
+                    IntStream.range(0, numReadsPerThread)
+                            .forEach(j -> results.addAll(readOneBatchOfSize(1)));
+                    return results;
+                })));
+        executorService.shutdown();
+        Preconditions.checkState(executorService.awaitTermination(30L, TimeUnit.SECONDS),
+                "Read tasks did not finish in 30s");
+        long timeTakenToReadInSeconds = TimeUnit.NANOSECONDS.toSeconds(System.nanoTime() - start);
```

Comment:
Note that this rounds things down to the nearest second, which could mess with the calculations later (i.e. 1,999,999,999 ns would become 1 s)